pkg/function/kubectl_test.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2021 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package function
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	. "gopkg.in/check.v1"
+	v1 "k8s.io/api/core/v1"
+	extensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	crdclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+
+	kanister "github.com/kanisterio/kanister/pkg"
+	crv1alpha1 "github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1"
+	"github.com/kanisterio/kanister/pkg/kube"
+	"github.com/kanisterio/kanister/pkg/param"
+)
+
+var _ = Suite(&KubectlSuite{})
+
+type KubectlSuite struct {
+	kubeCli   kubernetes.Interface
+	crdCli    crdclient.Interface
+	dynCli    dynamic.Interface
+	namespace string
+}
+
+func (s *KubectlSuite) SetUpSuite(c *C) {
+	cli, err := kube.NewClient()
+	c.Assert(err, IsNil)
+	s.kubeCli = cli
+
+	dynCli, err := kube.NewDynamicClient()
+	c.Assert(err, IsNil)
+	s.dynCli = dynCli
+
+	ns := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "kanisterkubetasktest-",
+		},
+	}
+	cns, err := s.kubeCli.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{})
+	c.Assert(err, IsNil)
+	s.namespace = cns.Name
+	// Create CRD
+	crdCli, err := kube.NewCRDClient()
+	c.Assert(err, IsNil)
+	s.crdCli = crdCli
+	_, err = s.crdCli.ApiextensionsV1().CustomResourceDefinitions().Create(context.TODO(), getSampleCRD(), metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(err) {
+		return
+	}
+	c.Assert(err, IsNil)
+}
+
+func (s *KubectlSuite) TearDownSuite(c *C) {
+	if s.namespace != "" {
+		_ = s.kubeCli.CoreV1().Namespaces().Delete(context.TODO(), s.namespace, metav1.DeleteOptions{})
+	}
+	_ = s.crdCli.ApiextensionsV1().CustomResourceDefinitions().Delete(context.TODO(), getSampleCRD().GetName(), metav1.DeleteOptions{})
+}
+
+func createPhase(namespace string) crv1alpha1.BlueprintPhase {
+	return crv1alpha1.BlueprintPhase{
+		Name: "create-in-ns",
+		Func: KubectlFuncName,
+		Args: map[string]interface{}{
+			KubectlOperationArg: "create",
+			KubectlNamespaceArg: namespace,
+			KubectlSpecsArg: fmt.Sprintf(`apiVersion: apps/v1
+kind: Deployment
```

Comment:
I would put this in `testdata` so you get yaml highlighting, etc.

Reply:
How about moving this to a raw string const?

Reply:
@tdmanv we also need to set namespace in the k8s object specs. It would be complicated if we want to read specs from a file and then covert it into k8s object and set namespace. For now, I am setting the specs as const.